### PR TITLE
Feature/Unit test in junit for io c

### PR DESCRIPTION
**Pull request by Ramchy-J**
- Created the unit test file for the IoC
- Update the IoC to work was intended based on the test

### CODE REVIEW

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+class IoCContainerTest {
+
+    @Test
+    void singletonIoCContainer() {
+
+        IoCContainer ioc = IoCContainer.getInstance();
+        IoCContainer ioc2 = IoCContainer.getInstance();
+
+        assertEquals(ioc, ioc2);
+    }
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: You don't need this test. Focus on testing what the code should do regarding your business logic

File: src/main/java/com/ot/restaurant/entities/IoCContainer.java [hunk 2]
```diff
@@ -34,7 +38,11 @@ public void register(final String key, final T value) throws Exception {
 
     Optional.ofNullable(key).orElseThrow(DependencyKeyNullpointerException::new);
 
-    Optional.of(key).map(instancePool::get).orElseThrow(DuplicatedDependencyFoundException::new);
+    //Optional.of(key).map(instancePool::get).orElseThrow(DuplicatedDependencyFoundException::new);
+
+    if (Optional.ofNullable(instancePool.get(key)).isPresent()) {
+      throw new DuplicatedDependencyFoundException();
+    }
```
Comment (reynoldmorel) [inline src/main/java/com/ot/restaurant/entities/IoCContainer.java:40]: Use `Optional.ifPresent(...)` instead

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Test
+    void testRegisterAndResolve() throws Exception {
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: Use a conventional naming for test such as:
```
void whenRegisterValueShouldResolveValue ()
```

The pattern is something similar to: `when_Should_` or `when_Then_` or `when_Must_`

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+        assertEquals(value, valueResolvedIoC);
+    }
+
+    @Test
+    void testDuplicated() throws Exception {
+        IoCContainer ioc = IoCContainer.getInstance();
+
+        //If we change the value of key2 from "keyEntiti2" to "keyEntiti1" it will throw the duplicated exception
+        //because the wash already registered by key1. or if we uncomment the second registration of key1.
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: No need to add comments here

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+        ioc.register(key, value);
+
+        Object valueResolvedIoC = ioc.resolve(key);
+
+        assertEquals(value, valueResolvedIoC);
+    }
+
+    @Test
+    void testDuplicated() throws Exception {
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: Change the name according to above comment

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        //If we change the value of key2 from "keyEntiti2" to "keyEntiti1" it will throw the duplicated exception
+        //because the wash already registered by key1. or if we uncomment the second registration of key1.
+        Object value1 = new Object();
+        Object value2 = new Object();
+        String key1 = "keyEntiti1";
+        String key2 = "keyEntiti2";
+
+        ioc.register(key1, value1);
+        ioc.register(key1, value1);
+        //ioc.register(key2, value2);
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: Assert exception here

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Test
+    void testRegisterNullpointerException() throws Exception {
+        IoCContainer ioc = IoCContainer.getInstance();
+
+        Object value1 = new Object();
+        String key1 = null;
+
+        ioc.register(key1, value1);
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: Assert exception here

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+
+    @Test
+    void testResolveNullpointerException() throws Exception {
+        IoCContainer ioc = IoCContainer.getInstance();
+
+        String key1 = null;
+
+        ioc.resolve(key1);
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: ditto

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    void testResolveNotFoundException() throws Exception {
+        IoCContainer ioc = IoCContainer.getInstance();
+
+        Object value1 = new Object();
+        String key1 = "keyEntiti1";
+        String key2 = "keyEntiti2";
+
+        ioc.register(key1, value1);
+
+        Object valueResolvedIoC = ioc.resolve(key2);
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: ditto

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 3 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 5 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 7 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 9 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 8 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 6 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 4 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 1 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  @Test
+  void whenRegisterThrowsNullExceptionThenAssertionSucceeds() {
+    IoCContainer ioc = IoCContainer.getInstance();
+
+    Object value1 = new Object();
+    String key1 = "Ramon";
+
+    Exception thrown =
+        Assertions.assertThrows(
+            Exception.class,
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java:65]: I think you should use the exception class you created

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+  @Test
+  void whenResolveThrowsNotFoundExceptionThenAssertionSucceeds() {
+    IoCContainer ioc = IoCContainer.getInstance();
+
+    Object value1 = new Object();
+
+    Exception thrown =
+        Assertions.assertThrows(
+            Exception.class,
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: I think you should use the exception class you created

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    Object value1 = new Object();
+
+    String key2 = "keyEntiti1";
+    Object value2 = new Object();
+
+    //    ioc.register(key1, value1);
+
+    Exception thrown =
+        Assertions.assertThrows(
+            Exception.class,
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: I think you should use the exception class you created

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 10 above)
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: I think you should use the exception class you created

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+    assertEquals(ioc, ioc2);
+  }
+
+  @Test
+  void whenRegisterValueShouldResolveValueThenAssertSucceeds() throws Exception {
+    IoCContainer ioc = IoCContainer.getInstance();
+
+    String key = "keyEntiti";
+    Object value = new Object();
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: Could you use the test setup to always setup a:
1. ioc
2. key
3. value

?

File: src/main/java/com/ot/restaurant/entities/IoCContainer.java (hunk 2 above)
Comment (reynoldmorel) [reply src/main/java/com/ot/restaurant/entities/IoCContainer.java:40]: What do you think about this:
```
Optional.ofNullable(instancePool.get(key).ifPresent((val) -> throw new DuplicatedDependencyFoundException())
```

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 12 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 10 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 11 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 13 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: Yes, Done

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    com.ot.restaurant.exceptions.DependencyNotFoundException thrown =
+        Assertions.assertThrows(
+            DependencyNotFoundException.class,
+            () -> {
+              String key = "keyEntiti1";
+
+              ioc.resolve(key);
+            });
+
+    Assertions.assertEquals("Dependency Value does not exists", thrown.getMessage());
```
Comment (reynoldmorel) [inline src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: You can remove this line. Asseting the exception thown is `DependencyNotFoundException` should be enough

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/test/java/com/ot/restaurant/entities/IoCContainerTest.java (hunk 14 above)
Comment (Ramchy-J) [reply src/test/java/com/ot/restaurant/entities/IoCContainerTest.java]: Done

**Review by reynoldmorel — APPROVED** [review]

**Pull request merged**